scripts/knot_benchmark/src/benchmark/cases.py

```diff
@@ -0,0 +1,223 @@
+
+    def cold_command(self, project: Project, venv: Venv) -> Command:
+        command = [str(self.path), "-v"]
+
+        assert len(project.include) < 2, "Knot doesn't support multiple source folders"
+
+        if directory := project.include[0]:
+            command.extend(["--current-directory", directory])
+
+        command.extend(["--venv-path", str(venv.path)])
+
+        return Command(
+            name="knot",
+            command=command,
+        )
+
+
+class MyPy(Tool):
+    path: Path
+
+    def __init__(self, *, path: Path | None = None):
+        self.path = path or which_tool(
+            "mypy",
+        )
+
+    def cold_command(self, project: Project, venv: Venv) -> Command:
+        command = [
+            *self._base_command(project, venv),
+            "--no-incremental",
+            "--cache-dir",
+            os.devnull,
+        ]
+
+        return Command(
+            name="mypy",
+            command=command,
+        )
+
+    def warm_command(self, project: Project, venv: Venv) -> Command | None:
+        command = [
+            str(self.path),
+            *(project.mypy_arguments or project.include),
+            "--python-executable",
+            str(venv.python),
+        ]
+
+        return Command(
+            name="mypy",
+            command=command,
+        )
+
+    def _base_command(self, project: Project, venv: Venv) -> list[str]:
+        return [
+            str(self.path),
+            "--python-executable",
+            str(venv.python),
+            *(project.mypy_arguments or project.include),
+        ]
+
+
+class Pyright(Tool):
+    path: Path
+
+    def __init__(self, *, path: Path | None = None):
+        self.path = path or which_tool("pyright")
+
+    def cold_command(self, project: Project, venv: Venv) -> Command:
+        command = [
+            str(self.path),
+            "--venvpath",
+            str(
+                venv.path.parent
+            ),  # This is not the path to the venv folder, but the folder that contains the venv...
+            *(project.pyright_arguments or project.include),
+        ]
+
+        return Command(
+            name="Pyright",
+            command=command,
+        )
+
+
+class Venv:
+    path: Path
+
+    def __init__(self, path: Path):
+        self.path = path
+
+    @property
+    def name(self):
+        """The name of the virtual environment directory."""
+        return self.path.name
+
+    @property
+    def python(self):
+        """Returns the path to the python executable"""
+        return self.script("python")
+
+    @property
+    def bin(self) -> Path:
+        bin_dir = "scripts" if sys.platform == "win32" else "bin"
+        return self.path.joinpath(bin_dir)
+
+    def script(self, name: str) -> Path:
+        extension = ".exe" if sys.platform == "win32" else ""
+        return self.bin.joinpath(f"{name}{extension}")
```

Comment:
it's generally considered more idiomatic to use the `/` operator when working with pathlib. I almost never see the `joinpath` method being used 😆

```suggestion
        return self.bin / "{name}{extension}"
```

Reply:
Yeah, not a huge fan of magic operator overloads. It gives the impression that these are numbers. That's why I intentionally used the method but I'm open to changing it if we find this more idiomatic

Reply:
Python developers will be accustomed to seeing `/` used when working with pathlib, and I think would mostly (like I was) be quite surprised to see `joinpath` being used throughout instead of `/`. But I know that not everybody on the team has a huge amount of experience with Python, so I guess there's a tradeoff there.